[Security Solution][Detections] Add bulk rule action for manual rule run (#9653)

e40pud opened the pull request:
Main ticket https://github.com/elastic/security-team/issues/9653

With this changes we introduce a new bulk action which allows to schedule backfill for multiple rules.

**NOTES**:
- To be able to test these changes, you need to enable feature flag `manualRuleRunEnabled` first

**RECORDING**:

https://github.com/elastic/kibana/assets/2700761/742083e7-090e-4805-8c3d-abcba04554b1

### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
    - [x] https://github.com/elastic/security-docs/issues/5264
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
    - [x] [Cypress RM (100 ESS & 100 Serverless)](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6410)
    - [x] [Cypress DE (100 ESS & 100 Serverless)](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6411)
    - [x] [Integration Rule Gaps (100 ESS & 100 Serverless)](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6412)
    - [x] [Integration Bulk Actions (100 ESS & 100 Serverless)](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6413)

e40pud committed:
[Security Solution][Detections] Add bulk rule action for manual rule run (#9653)

e40pud committed:
Fixing type check error

TS7056: The inferred type of this node exceeds the maximum length the compiler will serialize. An explicit type annotation is needed.

https://buildkite.com/elastic/kibana-pull-request/builds/216105

e40pud commented:
/ci

e40pud committed:
Fix types

e40pud commented:
/ci

e40pud committed:
Fix tests

kibanamachine committed:
Merge branch 'main' into security/feature/9653-bulk-backfill

e40pud commented:
/ci

e40pud commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into security/feature/9653-bulk-backfill

e40pud commented:
/ci

e40pud commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into security/feature/9653-bulk-backfill

e40pud commented:
/ci

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

elasticmachine commented:
Pinging @elastic/security-detection-engine (Team:Detection Engine)

kibanamachine commented:
## Flaky Test Runner Stats
### 🎉 All tests passed! - [kibana-flaky-test-suite-runner#6356](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6356)
[✅] x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_gaps/trial_license_complete_tier/configs/ess.config.ts: 100/100 tests passed.
[✅] x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_gaps/trial_license_complete_tier/configs/serverless.config.ts: 100/100 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F186293%2Fhead)

kibanamachine commented:
## Flaky Test Runner Stats
### 🎉 All tests passed! - [kibana-flaky-test-suite-runner#6357](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6357)
[✅] x-pack/test/security_solution_api_integration/test_suites/detections_response/rules_management/rule_bulk_actions/trial_license_complete_tier/configs/ess.config.ts: 100/100 tests passed.
[✅] x-pack/test/security_solution_api_integration/test_suites/detections_response/rules_management/rule_bulk_actions/trial_license_complete_tier/configs/serverless.config.ts: 100/100 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F186293%2Fhead)

kibanamachine commented:
## Flaky Test Runner Stats
### 🎉 All tests passed! - [kibana-flaky-test-suite-runner#6354](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6354)
[✅] Security Solution Rule Management - Cypress: 100/100 tests passed.
[✅] [Serverless] Security Solution Rule Management - Cypress: 100/100 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F186293%2Fhead)

kibanamachine commented:
## Flaky Test Runner Stats
### 🟠 Some tests failed. - [kibana-flaky-test-suite-runner#6355](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6355)
[✅] Security Solution Detection Engine - Cypress: 100/100 tests passed.
[❌] [Serverless] Security Solution Detection Engine - Cypress: 74/100 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F186293%2Fhead)

nkhristinin reviewed (changes requested):
Hey, great work! 

Tested locally, looks good!

I left some comments, mostly it's questions about some potenial duplicate code

And there some requested changes for rename Backfill to Manual run

e40pud commented on x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/bulk_actions/bulk_action_rule_errors_list.tsx:
```diff
@@ -108,6 +111,75 @@ const BulkExportRuleErrorItem = ({
+        <li key={message}>
+          <FormattedMessage
+            id="xpack.securitySolution.detectionEngine.rules.allRules.bulkActions.scheduleDisabledRuleDescription"
+            defaultMessage="{rulesCount, plural, =1 {# rule} other {# rules}} (Cannot schedule backfill for disabled rules)"
```
will update copies

e40pud commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into security/feature/9653-bulk-backfill

e40pud commented on x-pack/plugins/security_solution/public/detection_engine/rule_management/api/api.ts:
```diff
@@ -380,6 +412,35 @@ export async function performBulkAction({
+ *
+ * @throws An error if response is not OK
+ */
+export async function performBulkScheduleBackfillAction({
```
The `performBulkAction` method runs bulk actions which modify the rule (create, update, delete etc.) and returns the corresponding results object:

```
export interface BulkActionResult {
  updated: RuleResponse[];
  created: RuleResponse[];
  deleted: RuleResponse[];
  skipped: RuleResponse[];
}
```

At first, I tried to add a `backfilled` property to `BulkActionResult` and make `performBulkAction` a generic method which will cover backfill bulk actions, but then I decided not to go with this approach because:
1. I did not like the mixture of different unrelated actions. Right now `performBulkAction` handles everything in a way that it expects the `CRUD` actions. It reflected in types names and objects that we return.
2. When I tried to add `backfilled: RuleResponse[];` to the `BulkActionResult`, I've got an error about complexity of the type: `"The inferred type of this node exceeds the maximum length the compiler will serialize. An explicit type annotation is needed."`. To overcome it, we would need to re-structure  the whole `BulkActionResult` in a way to make it simpler for the typescript compiler and to make sure that we do not break current APIs.

jpdjere replied:
Did you consider here reusing "updated" instead of creating a new `backfilled` prop? Do you think it's just wrong from an API design point of view? (I understand your point about CRUD vs scheduling a run)

My point here is that reusing `performBulkAction` also means not having to create a separate `buildBulkScheduleBackfillResponse` or modify the types, or create new mutations at all. If the cost to pay for that is only having an API response that is not as specific as it should be, then maybe it's a cost we want to pay.

e40pud replied:
Updated to use `performBulkAction` for "manual rule run" action as well.

e40pud commented on x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/bulk_actions/validations.ts:
```diff
@@ -76,6 +86,68 @@ export const validateBulkDuplicateRule = async ({ rule, mlAuthz }: BulkActionsVa
+ * runs validation for bulk schedule backfill for a single rule
+ * @param params - {@link DryRunScheduleBackfillBulkActionsValidationArgs}
+ */
+export const validateBulkScheduleBackfill = async ({
```
Agree! Talked to @ymao1 and asked to update error schema that we return as a result of the schedule backfill call https://github.com/elastic/kibana/pull/186638. Once that is done, I will be able to simplify this logic and remove validations from security solution side.

e40pud replied:
done

jpdjere commented on x-pack/plugins/security_solution/public/detection_engine/rule_management/api/api.ts:
```diff
@@ -299,6 +300,7 @@ export interface BulkActionResult {
   created: RuleResponse[];
   deleted: RuleResponse[];
   skipped: RuleResponse[];
+  backfilled: RuleResponse[];
```
Looks like you extended this result type here with `backfilled` but ended up using `BulkScheduleBackfillActionResponse` directly, and `backfilled` is unused here.

e40pud replied:
done

jpdjere commented on x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/api/rules/bulk_actions/route.ts:
```diff
@@ -322,6 +323,23 @@ export const performBulkActionRoute = (
+              errors.push(...bulkActionErrors);
+              return buildBulkScheduleBackfillResponse(response, {
+                errors,
+                isDryRun,
```
Building the response seems to be the same whether it is `dryRun` or not.

e40pud replied:
done

jpdjere commented on x-pack/plugins/security_solution/public/detection_engine/rule_management/api/hooks/use_bulk_schedule_backfill_mutation.ts:
```diff
@@ -0,0 +1,39 @@
+} from '../api';
+import { performBulkScheduleBackfillAction } from '../api';
+
+export const BULK_ACTION_MUTATION_KEY = ['POST', DETECTION_ENGINE_RULES_BULK_ACTION];
```
If we are keeping this mutation hook separate from `useBulkActionMutation`, we should modify the `BULK_ACTION_MUTATION_KEY` value that we use for the mutation key to avoid unwanted caching invalidations with the other hook, which uses exactly the same key.

I think something like:
```ts
export const BULK_SCHEDULE_BACKFILL_MUTATION_KEY = ['POST', DETECTION_ENGINE_RULES_BULK_ACTION, 'SCHEDULE_BACKFILL'];
```
should suffice.

e40pud replied:
removed this mutation hook

jpdjere commented on x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/bulk_actions/use_bulk_actions.tsx:
```diff
@@ -53,19 +57,25 @@ interface UseBulkActionsArgs {
     action: BulkActionForConfirmation
   ) => Promise<boolean>;
   showBulkDuplicateConfirmation: () => Promise<string | null>;
+  showManualRuleRunConfirmation: () => Promise<TimeRange | null>;
+  showScheduleBackfillLimitError: () => void;
```
Any reason why these two props use different terminology? ManualRun vs Backfill.

If possible I'd keep them unified in the code as well. My personal preference is for "Manual Run" instead of "Backfill" since it seems more self-explanatory by itself, but I understand that Backfill is maybe the technical term - and it's only one word.

e40pud committed:
Merge branch 'main' into security/feature/9653-bulk-backfill

e40pud commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into security/feature/9653-bulk-backfill

e40pud committed:
Review feedback

- use "bulk edit action" path for manual rule run action
- adjust error handling to new backfill API error format
- update messages to use "manual rule run" instead of "schedule backfile"

e40pud committed:
Remove redundant changes

e40pud committed:
Remove redundant changes

e40pud committed:
Remove redundant changes

e40pud committed:
Update error messages

e40pud committed:
Update cypress tests

e40pud commented on x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/bulk_actions/bulk_action_rule_errors_list.tsx:
```diff
@@ -108,6 +111,75 @@ const BulkExportRuleErrorItem = ({
+        <li key={message}>
+          <FormattedMessage
+            id="xpack.securitySolution.detectionEngine.rules.allRules.bulkActions.scheduleFarInPastRuleDescription"
+            defaultMessage="{rulesCount, plural, =1 {# rule} other {# rules}} (Backfill cannot look back more than {daysCount, plural, =1 {# day} other {# days}})"
```
done

e40pud commented on x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/bulk_actions/bulk_action_rule_errors_list.tsx:
```diff
@@ -108,6 +111,75 @@ const BulkExportRuleErrorItem = ({
+        <li key={message}>
+          <FormattedMessage
+            id="xpack.securitySolution.detectionEngine.rules.allRules.bulkActions.scheduleInvalidTimeRangeRuleDescription"
+            defaultMessage="{rulesCount, plural, =1 {# rule} other {# rules}} (Backfill end must be greater than backfill start)"
```
done

e40pud commented on x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/api/rules/bulk_actions/bulk_actions_response.ts:
```diff
@@ -100,6 +103,55 @@ export const buildBulkResponse = (
+export const buildBulkScheduleBackfillResponse = (
+  response: KibanaResponseFactory,
+  {
+    isDryRun = false,
```
done

e40pud commented on x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/bulk_actions/bulk_action_rule_errors_list.tsx:
```diff
@@ -108,6 +111,75 @@ const BulkExportRuleErrorItem = ({
+        <li key={message}>
+          <FormattedMessage
+            id="xpack.securitySolution.detectionEngine.rules.allRules.bulkActions.scheduleInFutureRuleDescription"
+            defaultMessage="{rulesCount, plural, =1 {# rule} other {# rules}} (Backfill cannot be scheduled for the future)"
```
done

kibanamachine commented:
## Flaky Test Runner Stats
### 🎉 All tests passed! - [kibana-flaky-test-suite-runner#6412](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6412)
[✅] x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_gaps/trial_license_complete_tier/configs/ess.config.ts: 100/100 tests passed.
[✅] x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_gaps/trial_license_complete_tier/configs/serverless.config.ts: 100/100 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F186293%2Fhead)

nkhristinin reviewed (approved):
LGTM, this is great feature!

kibanamachine commented:
## Flaky Test Runner Stats
### 🎉 All tests passed! - [kibana-flaky-test-suite-runner#6413](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6413)
[✅] x-pack/test/security_solution_api_integration/test_suites/detections_response/rules_management/rule_bulk_actions/trial_license_complete_tier/configs/ess.config.ts: 100/100 tests passed.
[✅] x-pack/test/security_solution_api_integration/test_suites/detections_response/rules_management/rule_bulk_actions/trial_license_complete_tier/configs/serverless.config.ts: 100/100 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F186293%2Fhead)

jpdjere commented on x-pack/plugins/security_solution/public/common/lib/apm/user_actions.ts:
```diff
@@ -23,6 +23,7 @@ export const BULK_RULE_ACTIONS = {
   DISABLE: `${APP_UI_ID} bulkRuleActions disable`,
   DUPLICATE: `${APP_UI_ID} bulkRuleActions duplicate`,
   EXPORT: `${APP_UI_ID} bulkRuleActions export`,
+  SCHEDULE_BACKFILL: `${APP_UI_ID} bulkRuleActions schedule backfill`,
```
nit: this is the only remaining place where we use the term "backfil"

e40pud replied:
done

jpdjere commented on x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/bulk_actions/use_bulk_actions.tsx:
```diff
@@ -202,6 +208,53 @@ export const useBulkActions = ({
       };
 
+      const handleScheduleRuleRunAction = async () => {
+        startTransaction({ name: BULK_RULE_ACTIONS.SCHEDULE_BACKFILL });
```
nit: and also here

jpdjere commented on x-pack/test/security_solution_cypress/cypress/e2e/detection_response/detection_engine/rule_gaps/bulk_manual_rule_run.cy.ts:
```diff
@@ -0,0 +1,79 @@
+    cy.contains(TOASTER_BODY, `Successfully scheduled manual rule run for ${enabledCount} rule`);
+  });
+
+  it('schedule enable rules and show partial error for siabled rules when all rules are selected', () => {
```
`siabled` -> `disabled`

e40pud replied:
done

jpdjere commented on x-pack/test/security_solution_cypress/cypress/e2e/detection_response/detection_engine/rule_gaps/bulk_manual_rule_run.cy.ts:
```diff
@@ -0,0 +1,79 @@
+    scheduleManualRuleRunForSelectedRules(enabledCount, disabledCount);
+
+    cy.contains(TOASTER_BODY, `Successfully scheduled manual rule run for ${enabledCount} rule`);
+  });
```
Are we missing an assertion here that checks that there's a warning about disabled rules?

e40pud replied:
It is done within `scheduleManualRuleRunForSelectedRules` method. We pass `disabledCount` and if it is > 0 thats we show and handle warning modal.

jpdjere reviewed (approved):
LGTM! 🚀 Thanks for working on this, hugely desired feature.

Thanks as well for implementing our suggestions on reusing `performBulkAction`, I think the PR looks much more concise now and there's much less new code.

Left a couple nits and typo warnings, but 👍 💯

e40pud committed:
Review feedback

kibanamachine commented:
## Flaky Test Runner Stats
### 🎉 All tests passed! - [kibana-flaky-test-suite-runner#6410](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6410)
[✅] Security Solution Rule Management - Cypress: 100/100 tests passed.
[✅] [Serverless] Security Solution Rule Management - Cypress: 100/100 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F186293%2Fhead)

kibanamachine commented:
## Flaky Test Runner Stats
### 🟠 Some tests failed. - [kibana-flaky-test-suite-runner#6411](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6411)
[✅] Security Solution Detection Engine - Cypress: 100/100 tests passed.
[❌] [Serverless] Security Solution Detection Engine - Cypress: 75/100 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F186293%2Fhead)
